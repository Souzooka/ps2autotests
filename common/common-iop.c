#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
