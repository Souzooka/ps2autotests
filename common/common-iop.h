#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
