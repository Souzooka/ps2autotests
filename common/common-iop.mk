# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
