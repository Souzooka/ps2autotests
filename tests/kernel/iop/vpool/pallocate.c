#include <common-iop.h>
#include <thpool.h>
#include "vpl-common.h"

#define MAX_ALLOC_SIZES 24

static const s32 g_allocSizes[MAX_ALLOC_SIZES] = {
	0x80, 0x81, 0x82, 0x83, 
	0x84, 0x85, 0x86, 0x87, 
	0x88, 0x89, 0x8A, 0x8B,
	0x8C, 0x8D, 0x8E, 0x8F,
	0x100, 0x800, 0x1000, 0x8000,
	-0x1, -0x5, -0x8, -0x80
};

void doAllocate(u32 vplId, s32 size) {
	void *ptr = pAllocateVpl(vplId, size);
	s32 result = (s32)ptr;
	if(result >= 0) {
		printf("  succeeded -> ");
		printVpl(vplId);
		FreeVpl(vplId, ptr);
	} else {
		printf("  failed -> result (%d)\n", result);
	}
} 

void testSingleAllocs() {
	iop_vpl_param vplParam = {};
	vplParam.attr   = VA_THFIFO;
	vplParam.option = 0;
	vplParam.size   = 0x1000;
	
	s32 vplId = CreateVpl(&vplParam);
	
	s32 i = 0;
	for(i = 0; i < MAX_ALLOC_SIZES; i++) {
		s32 allocSize = g_allocSizes[i];
		printf("  allocate %d bytes:\n", allocSize);
		printf("  "); doAllocate(vplId, allocSize);
	}
	
	DeleteVpl(vplId);
}

void testMultiAllocs() {
	iop_vpl_param vplParam = {};
	vplParam.attr   = VA_THFIFO;
	vplParam.option = 0;
	vplParam.size   = 0x1000;
	
	s32 vplId = CreateVpl(&vplParam);
	
	pAllocateVpl(vplId, 0x800);
	pAllocateVpl(vplId, 0x400);
	pAllocateVpl(vplId, 0x200);
	pAllocateVpl(vplId, 0x100);
	
	s32 i = 0;
	for(i = 0; i < MAX_ALLOC_SIZES; i++) {
		s32 allocSize = g_allocSizes[i];
		printf("  allocate %d bytes:\n", allocSize);
		printf("  "); doAllocate(vplId, allocSize);
	}
	
	DeleteVpl(vplId);
}

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	printf("single allocations:\n");
	testSingleAllocs();
	
	printf("multiple allocations:\n");
	testMultiAllocs();
	
	printf("alloc from invalid vpl:\n");
	doAllocate(~0, 0x100);
	
	printf("alloc negative from invalid vpl:\n");
	doAllocate(~0, -0x80);
	
	printf("-- TEST END\n");
	return 0;
}
