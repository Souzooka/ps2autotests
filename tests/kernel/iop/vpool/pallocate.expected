-- TEST BEGIN
single allocations:
  allocate 128 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3920
  allocate 129 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3912
  allocate 130 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3912
  allocate 131 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3912
  allocate 132 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3912
  allocate 133 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3912
  allocate 134 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3912
  allocate 135 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3912
  allocate 136 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3912
  allocate 137 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3904
  allocate 138 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3904
  allocate 139 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3904
  allocate 140 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3904
  allocate 141 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3904
  allocate 142 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3904
  allocate 143 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3904
  allocate 256 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 3792
  allocate 2048 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 2000
  allocate 4096 bytes:
    failed -> result (-427)
  allocate 32768 bytes:
    failed -> result (-427)
  allocate -1 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 4048
  allocate -5 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 4048
  allocate -8 bytes:
    failed -> result (-400)
  allocate -128 bytes:
    failed -> result (-400)
multiple allocations:
  allocate 128 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 48
  allocate 129 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 40
  allocate 130 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 40
  allocate 131 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 40
  allocate 132 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 40
  allocate 133 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 40
  allocate 134 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 40
  allocate 135 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 40
  allocate 136 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 40
  allocate 137 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 32
  allocate 138 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 32
  allocate 139 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 32
  allocate 140 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 32
  allocate 141 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 32
  allocate 142 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 32
  allocate 143 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 32
  allocate 256 bytes:
    failed -> result (-400)
  allocate 2048 bytes:
    failed -> result (-400)
  allocate 4096 bytes:
    failed -> result (-427)
  allocate 32768 bytes:
    failed -> result (-427)
  allocate -1 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 176
  allocate -5 bytes:
    succeeded -> attr: 0, option: 0, wait threads: 0, size: 4056, free size: 176
  allocate -8 bytes:
    failed -> result (-400)
  allocate -128 bytes:
    failed -> result (-400)
alloc from invalid vpl:
  failed -> result (-411)
alloc negative from invalid vpl:
  failed -> result (-411)
-- TEST END
